### PR TITLE
Context Relevancy v2

### PR DESCRIPTION
**Pull request by shahules786**
## What
Added support for new context relevancy metrics
## Why
The earlier used context_relevancy metric was found to have a poor correlation with actual results. 
## How
New paradigm leverages candidate sentence extraction with self-consistency checks to ensure far better results even with larger context lengths. 
## Screenshots 
Screenshot 2023-07-13 at 10 47 52 AM

### CODE REVIEW

File: src/ragas/metrics/context_relevance.py [hunk 1]
```diff
@@ -2,62 +2,137 @@
...
 
     def score(self: t.Self, dataset: Dataset) -> Dataset:
         """
         dataset: Dataset[question: list[str], contexts: list[list[str]]]
         """
+        prompts = []
+        questions, contexts = dataset["question"], dataset["contexts"]
+        for q, c in zip(questions, contexts):
+            prompt = CONTEXT_RELEVANCE.format(q, "\n".join(c))
+            prompts.append(prompt)
```
Comment (jjmachan) [inline src/ragas/metrics/context_relevance.py:104]: wouldn't this need batching. If there are more than 20 (questions, contexts) pairs openAI will return an error?

File: src/ragas/metrics/context_relevance.py [hunk 2]
```diff
@@ -2,62 +2,137 @@
...
+                ]
+                overlap_scores.append(len(indices) / len(context_sents))
+            if self.strictness > 1:
+                agr_score = self.sent_agreement.evaluate(n_output)
+            else:
+                agr_score = 1
+            scores.append(agr_score * np.mean(overlap_scores))
 
-        return score_ds
+        return dataset.add_column(f"{self.name}", scores)  # type: ignore
```
Comment (jjmachan) [inline src/ragas/metrics/context_relevance.py:135]: why are we type ignoring here? just curious

File: src/ragas/metrics/context_relevance.py [hunk 3]
```diff
@@ -2,62 +2,137 @@
...
 
 
 @dataclass
 class ContextRelevancy(Metric):
-    batch_size: int = 32
     name: str = "context_relavency"
-    model_name: str = "t5-base"
+    batch_size: int = 15
+    agreement_metric: str = "bert_score"
+    strictness: int = 2
```
Comment (jjmachan) [inline src/ragas/metrics/context_relevance.py:88]: what does the strictness param do? would be great to have doc string at this point I guess

File: src/ragas/metrics/context_relevance.py (hunk 2 above)
Comment (shahules786) [reply src/ragas/metrics/context_relevance.py:149]: A warning for not passing the dataset fingerprint argument, The same is done in the Faithfulness class.

**Comment by shahules786** [comment]
@jjmachan  Added the required changes.

**Review by jjmachan — APPROVED** [review]

**Pull request merged**